pkg/localstorev2/internal/upload/uploadstore.go

```diff
@@ -487,16 +492,77 @@ func (p *pushReporter) Report(
 		}
 
 		ui.Synced = now().Unix()
-		err = p.s.Store().Put(ui)
+		err = p.s.IndexStore().Put(ui)
 		if err != nil {
 			return fmt.Errorf("failed updating uploadItem %s: %w", ui, err)
 		}
 	}
 
-	err = p.s.Store().Put(ti)
+	err = p.s.IndexStore().Put(ti)
 	if err != nil {
 		return fmt.Errorf("failed updating tag: %w", err)
 	}
 
 	return nil
 }
+
+var (
+	errNextTagIDUnmarshalInvalidSize = errors.New("unmarshal nextTagID: invalid size")
+)
+
+// nextTagID is a storage.Item which stores a uint64 value in the store.
+type nextTagID uint64
+
+func (nextTagID) Namespace() string { return "upload" }
+
+func (nextTagID) ID() string { return "nextTagID" }
+
+func (n nextTagID) Marshal() ([]byte, error) {
```

Comment:
If renamed to MarshalBinary it will implement binary.BinaryMarshaler.

Reply:
@mrekucci @janos 

So currently the serialization interface is defined in the storagev2 package. This is a custom interface that is minimal in order to allow other types of serialization formats. For eg. json, protobuf etc. Going ahead we plan to use this same store as the statestore. There we will have other encodings.

I think, in future when we start using generics there will be a better solution to this. We can have a generic type that could implement different encodings and this way we can reuse the serialization. Wdyt?

Also, in order to change this we will have to change a lot of files. So either way, it would be better to do this in a separate PR.